codecs/av1_packet.go

```diff
@@ -0,0 +1,138 @@
+package codecs
+
+import (
+	"github.com/pion/rtp/v2/pkg/obu"
+)
+
+const (
+	zMask     = byte(0b10000000)
+	zBitshift = 7
+
+	yMask     = byte(0b01000000)
+	yBitshift = 6
+
+	wMask     = byte(0b00110000)
+	wBitshift = 4
+
+	nMask     = byte(0b00001000)
+	nBitshift = 3
+)
+
+// AV1Payloader payloads VP8 packets
+type AV1Payloader struct{}
+
+// Payload fragments a AV1 packet across one or more byte arrays
+func (p *AV1Payloader) Payload(mtu uint16, payload []byte) [][]byte {
+	return nil
+}
+
+// AV1Packet represents a depacketized AV1 RTP Packet
+//
+//  0 1 2 3 4 5 6 7
+// +-+-+-+-+-+-+-+-+
+// |Z|Y| W |N|-|-|-|
+// +-+-+-+-+-+-+-+-+
+//
+// https://aomediacodec.github.io/av1-rtp-spec/#44-av1-aggregation-header
+type AV1Packet struct {
+	// Z: MUST be set to 1 if the first OBU element is an
+	//    OBU fragment that is a continuation of an OBU fragment
+	//    from the previous packet, and MUST be set to 0 otherwise.
+	Z byte
+
+	// Y: MUST be set to 1 if the last OBU element is an OBU fragment
+	//    that will continue in the next packet, and MUST be set to 0 otherwise.
+	Y byte
+
+	// W: two bit field that describes the number of OBU elements in the packet.
+	//    This field MUST be set equal to 0 or equal to the number of OBU elements
+	//    contained in the packet. If set to 0, each OBU element MUST be preceded by
+	//    a length field. If not set to 0 (i.e., W = 1, 2 or 3) the last OBU element
+	//    MUST NOT be preceded by a length field. Instead, the length of the last OBU
+	//    element contained in the packet can be calculated as follows:
+	// Length of the last OBU element =
+	//    length of the RTP payload
+	//  - length of aggregation header
+	//  - length of previous OBU elements including length fields
+	W byte
+
+	// N: MUST be set to 1 if the packet is the first packet of a coded video sequence, and MUST be set to 0 otherwise.
+	N byte
+
+	OBUs [][]byte
+
+	// Buffer for fragmented OBU. If Unmarshal is called on a RTP Packet
+	// that doesn't contain a fully formed OBU
+	obuBuffer []byte
```

Comment:
I'm not sure if we should assume that the same instance of AV1packet can be reused for the subsequent RTP packets.  Users can see Z and Y flags. We can simply leave the fragment in the first and the last OBUs array IMO.

Reply:
If that you think is ugly, we could ditch the packet-by-packet "Unmarshal" semantics and use a callback style, e.g. OnOBUElement(func([]byte) error)... I may like it better IMHO.

Reply:
@enobufs So one nice thing about keeping `Unmarshal` is that it satisfies the `rtp.Depacketizer` interface. If I knew everything I know now I would have designed it differently :/

We assume that users keep the same H264 Packet across Unmarshals and has seemed to work decently well. I think we should make it more robust though. Checking things like SequenceNumbers etc...

For a new version of `pion/rtp` it would be great to completely rethinking how unmarshaling video works. Maybe something like `OnFrame` so it could work for all codecs?

Reply:
I'm rather against the callback style compared to the current unmarshal pattern. imo the callback style is not very go-thonic.

Re: rethinking how unmarshal works, @Sean-Der I have done a ReadFrame() pattern that pulls frames before. Basically you can pass in a byte reader and it returns a frame reader so something like:

```
type RTPReader interface {
  ReadRTP() (*rtp.Packet, error)
}

type FrameReader interface {
  ReadFrame() (*Frame, error)
}
```

then the unmarshaller:

```
DepacketizeH264(RTPReader) FrameReader
```

This allows on-demand-style unmarshalling, matches stdlib, and largely the motivation behind the rtpio package was to facilitate this sort of pattern for higher level objects.

I'm not against requiring the reuse of a packet type though (ie keep the current pattern). I think it's not that unreasonable.

Reply:
I love that! What if we added a `AV1Frame` `VP9Frame` with Unmarshal methods @enobufs @kevmo314 ?

then we can get both RTP payload awareness and higher level ‘I just want a video frame’ API

Reply:
Will update the PR with that today

Reply:
Sounds good to me! Yeah if we go through that route (and assuming all the codecs do), we could actually remove `SampleBuilder` completely, since all the codecs would be able to provide sample construction through an API.

And for some wishful thinking we can even pipe this all the way to `io.Writer` so depacketizing and writing to a file could be as simple as:

```
go io.Copy(f, rtp.NewIVFWriter(rtp.ParseAV1(av1track)))
```

😄

Reply:
@kevmo314 @enobufs What do you think of that?

I also see us keeping around the SampleBuilder? In the future it will only be concerned with filling holes/handling jitter? Since that code is codec agnostic we don't want to put it in a specific RTP codec.

Reply:
The read frame semantics is good and I agree it's better than callback style.
I realize I did not communicate my point very well. It was about `obuBuffer` being inside AV1Packet.

I can see two layers in this context (high to low):
* OBU Frame construction (Inside the IVFWriter example)
* RTP Payload parsing for AV1 (AV1Packet)

I believe the `obuBuffer` belongs to the "OBU Frame construction".

When I was trying to add the following code to IVFWrite to handle cases for missing RTP packets, it became clear that there needs to be a way to invalidate `obuBuffer` but it's not possible because it is a private member. I feel obuBuffer does not belong there. If I understood the above plan correctly, it should be in the FrameReader IMO.

AV1 support by Pion is super exciting!!